feat: Implement Airflow DAG for Amadeus ticket data ingestion

ykykyk0814 opened the pull request:
- Add ticket_data_ingestion DAG with 3 tasks: fetch, transform, load
- Integrate Amadeus API for flight offers search
- Fix date validation issue (use future dates instead of past)
- Implement data transformation from Amadeus format to internal schema
- Add database insertion with upsert logic
- Configure proper error handling and logging
- Test successful with 2 flight offers retrieved (SYD->BKK route)

Technical details:
- Uses OAuth2 authentication for Amadeus API
- Dynamic future date calculation (90 days ahead)
- PostgreSQL integration with conflict resolution
- Comprehensive error handling and retry logic

<img width="768" height="410" alt="image" src="https://github.com/user-attachments/assets/a99e58ed-f1f5-45c0-ae4f-b5d99e255d95" />

ykykyk0814 committed:
feat: Implement Airflow DAG for Amadeus ticket data ingestion

- Add ticket_data_ingestion DAG with 3 tasks: fetch, transform, load
- Integrate Amadeus API for flight offers search
- Fix date validation issue (use future dates instead of past)
- Implement data transformation from Amadeus format to internal schema
- Add database insertion with upsert logic
- Configure proper error handling and logging
- Test successful with 2 flight offers retrieved (SYD->BKK route)

Technical details:
- Uses OAuth2 authentication for Amadeus API
- Dynamic future date calculation (90 days ahead)
- PostgreSQL integration with conflict resolution
- Comprehensive error handling and retry logic

ykykyk0814 committed:
feat: Add comprehensive data validation and error handling

- Add custom exception classes (DataValidationError, AmadeusAPIError, DatabaseError)
- Implement validate_flight_data() function with comprehensive data quality checks
- Add send_alert() function for monitoring and alerting
- Enhanced error handling with specific error types and alerting
- Validate required fields, airport codes, timestamps, and pricing data
- Graceful handling of invalid data with logging and alerts
- Database error handling with proper exception propagation

Improves AIRFLOW-003 requirements:
- Data validation steps added to DAG
- Error handling and retries enhanced
- Alerting framework implemented (ready for email/Slack integration)

multiverse-exploration-kitten commented on airflow/dags/ticket_data_ingestion_dag.py:
```diff
@@ -0,0 +1,291 @@
+AMADEUS_CLIENT_ID = "fkJvYXyIa5OYqfK1ueh4HMaTdriW9g8z"
+AMADEUS_CLIENT_SECRET = "MAmbtUdvrUQe00Zq"
+AMADEUS_TOKEN_URL = "https://test.api.amadeus.com/v1/security/oauth2/token"
+AMADEUS_FLIGHT_OFFERS_URL = "https://test.api.amadeus.com/v2/shopping/flight-offers"
```
这些 key 永远不要commit 到github上，要放在.env里

multiverse-exploration-kitten commented on airflow/logs/dag_processor/latest:
logs就不用放上来了

multiverse-exploration-kitten commented on airflow/dags/ticket_data_ingestion_dag.py:
```diff
@@ -0,0 +1,291 @@
+                        error_msg = f"Error inserting record: {record}\n{e}"
+                        logging.error(error_msg)
+                        send_alert(error_msg, "ERROR")
+                        raise DatabaseError(f"Database insertion failed: {e}")
```
不要写这种散装逻辑。 把他refactor到一个dedicated class里

multiverse-exploration-kitten commented on airflow/dags/ticket_data_ingestion_dag.py:
```diff
@@ -0,0 +1,291 @@
+                        "departure_time": first_leg["departure"]["at"],
+                        "arrival_time": last_leg["arrival"]["at"],
+                        "seat_number": None,  # Amadeus does not provide seat info in offers
+                        "notes": offer.get("id"),
```
same here

ykykyk0814 committed:
refactor: improve code organization and security

- Move Amadeus API credentials to environment variables
- Create dedicated FlightTicketDatabaseHandler class for database operations
- Create dedicated FlightTicketTransformer class for data transformation
- Add comprehensive error handling and logging
- Update .gitignore to exclude Airflow logs
- Update README with setup instructions for Amadeus credentials
- Update docker-compose.airflow.yml to support environment variables

This refactoring improves:
- Security: No hardcoded credentials in version control
- Maintainability: Separated concerns into dedicated classes
- Testability: Easy to unit test individual components
- Error handling: Graceful handling of individual record failures
- Monitoring: Detailed logging and statistics

ykykyk0814 committed:
cleanup: remove Airflow logs directory

- Remove airflow/logs/ directory and all its contents
- Logs are now excluded from version control via .gitignore
- New logs will be created automatically when Airflow runs
- Keeps repository clean while maintaining functionality

ykykyk0814 committed:
delete logs
